chain/jsonrpc/src/lib.rs

```diff
@@ -1310,6 +1347,18 @@ fn status_handler(
     response.boxed()
 }
 
+fn debugz_handler(
+    handler: web::Data<JsonRpcHandler>,
+) -> impl Future<Output = Result<HttpResponse, HttpError>> {
+    let response = async move {
+        match handler.debugz().await {
+            Ok(value) => Ok(HttpResponse::Ok().json(&value)),
+            Err(_) => Ok(HttpResponse::MethodNotAllowed().finish()),
+        }
+    };
+    response.boxed()
+}
```

Comment:
I wonder if this can be just

```suggestion
async fn debugz_handler(
    handler: web::Data<JsonRpcHandler>,
) -> Result<HttpResponse, HttpError> {
        match handler.debugz().await {
            Ok(value) => Ok(HttpResponse::Ok().json(&value)),
            Err(_) => Ok(HttpResponse::MethodNotAllowed().finish()),
        }
}
```

?

Can't check immediatelly -- waiting for nearcore to compile on my trusty old laptop :) Already miss my 12 core machine :(

Reply:
Compiled!

So, yeah, this seems to work

```diff
diff --git a/chain/jsonrpc/src/lib.rs b/chain/jsonrpc/src/lib.rs
index dcbe658f1..92c00b0c3 100644
--- a/chain/jsonrpc/src/lib.rs
+++ b/chain/jsonrpc/src/lib.rs
@@ -1347,16 +1347,13 @@ fn status_handler(
     response.boxed()
 }
 
-fn debugz_handler(
+async fn debugz_handler(
     handler: web::Data<JsonRpcHandler>,
-) -> impl Future<Output = Result<HttpResponse, HttpError>> {
-    let response = async move {
-        match handler.debugz().await {
-            Ok(value) => Ok(HttpResponse::Ok().json(&value)),
-            Err(_) => Ok(HttpResponse::MethodNotAllowed().finish()),
-        }
-    };
-    response.boxed()
+) -> Result<HttpResponse, HttpError> {
+    match handler.debugz().await {
+        Ok(value) => Ok(HttpResponse::Ok().json(&value)),
+        Err(_) => Ok(HttpResponse::MethodNotAllowed().finish()),
+    }
 }
 
 fn health_handler(
```

While working on, it I've noticed another thing: we treat any error as debug being disabled by configuration. That is, if gettind debug info fails at runtime, the error will be silently swallowed. We'll see a 400, not a 500. To fix *that*, I'd do

```diff
diff --git a/chain/jsonrpc/src/lib.rs b/chain/jsonrpc/src/lib.rs
index dcbe658f1..7d77942d4 100644
--- a/chain/jsonrpc/src/lib.rs
+++ b/chain/jsonrpc/src/lib.rs
@@ -897,20 +897,17 @@ impl JsonRpcHandler {
     pub async fn debugz(
         &self,
     ) -> Result<
-        near_jsonrpc_primitives::types::status::RpcStatusResponse,
+        Option<near_jsonrpc_primitives::types::status::RpcStatusResponse>,
         near_jsonrpc_primitives::types::status::RpcStatusError,
     > {
         if self.enable_debug_rpc {
-            Ok(self
-                .client_addr
-                .send(Status { is_health_check: false, get_detailed_info: true })
-                .await??
-                .into())
-        } else {
-            Err(near_jsonrpc_primitives::types::status::RpcStatusError::InternalError {
-                error_message: "Debug not enabled in configuration.".to_string(),
-            })
+            return None;
         }
+        let status_response = self
+            .client_addr
+            .send(Status { is_health_check: false, get_detailed_info: true })
+            .await??;
+        Ok(Some(status_response.into()))
     }
 
     /// Expose Genesis Config (with internal Runtime Config) without state records to keep the
@@ -1347,16 +1344,12 @@ fn status_handler(
     response.boxed()
 }
 
-fn debugz_handler(
-    handler: web::Data<JsonRpcHandler>,
-) -> impl Future<Output = Result<HttpResponse, HttpError>> {
-    let response = async move {
-        match handler.debugz().await {
-            Ok(value) => Ok(HttpResponse::Ok().json(&value)),
-            Err(_) => Ok(HttpResponse::MethodNotAllowed().finish()),
-        }
-    };
-    response.boxed()
+async fn debugz_handler(handler: web::Data<JsonRpcHandler>) -> Result<HttpResponse, HttpError> {
+    match handler.debugz().await {
+        Ok(Some(value)) => Ok(HttpResponse::Ok().json(&value)),
+        Ok(None) => Ok(HttpResponse::MethodNotAllowed().finish()),
+        Err(_) => Ok(HttpResponse::ServiceUnavailable().finish()),
+    }
 }
 
 fn health_handler(
```

Reply:
done